File: scripts/market-brief-runner/python/run_market_brief_runner.py
from __future__ import annotations

import json
import os
import sys
import urllib.error
import urllib.request
from typing import Any

from market_brief_writer import build_market_brief_payload
from market_data_sources import fetch_a_share_market_snapshot, has_core_market_data


def main() -> int:
    load_dotenv_if_available()

    base_url = normalize_base_url(os.getenv("WORKSTATION_BASE_URL"))
    runner_secret = os.getenv("MARKET_BRIEF_RUNNER_SECRET")
    market = os.getenv("MARKET_BRIEF_MARKET", "A股")
    runner_name = os.getenv("MARKET_BRIEF_RUNNER_NAME", "akshare-runner")
    data_mode = os.getenv("MARKET_BRIEF_DATA_MODE", "akshare")

    if not base_url:
        print("WORKSTATION_BASE_URL is required, for example http://localhost:3000", file=sys.stderr)
        return 1

    if not runner_secret:
        print("MARKET_BRIEF_RUNNER_SECRET is required.", file=sys.stderr)
        return 1

    job: dict[str, Any] | None = None

    try:
        claim = post_json(
            base_url,
            "/api/market-briefs/skill-jobs/claim",
            {"market": market, "runner_name": runner_name},
            runner_secret,
        )
        if not claim.get("job_id"):
            print(claim.get("message") or "No queued market brief generation jobs.")
            return 0

        job = claim
        print(f"Claimed market brief job {job['job_id']} for {job['market']} {job['brief_date']}.")

        snapshot = fetch_a_share_market_snapshot(
            brief_date=job["brief_date"],
            market=job.get("market") or market,
            runner_name=runner_name,
            data_mode=data_mode,
        )

        if not has_core_market_data(snapshot):
            warnings = snapshot.get("meta", {}).get("warnings") or []
            message = "Real market data unavailable; no core index, breadth, or sector data fetched."
            if warnings:
                message = f"{message} Warnings: {'; '.join(warnings[:3])}"
            fail_job(base_url, runner_secret, job["job_id"], message)
            print(message, file=sys.stderr)
            return 1

        result_payload = build_market_brief_payload(job, snapshot)
        result = post_json(base_url, "/api/market-briefs/skill-result", result_payload, runner_secret)
        print(f"Market brief generated: {result.get('preview_url')}")
        return 0
    except Exception as exc:  # noqa: BLE001
        message = safe_error(exc)
        if job:
            fail_job(base_url, runner_secret, job["job_id"], message)
        print(message, file=sys.stderr)
        return 1


def post_json(base_url: str, path: str, payload: dict[str, Any], runner_secret: str) -> dict[str, Any]:
    request = urllib.request.Request(
        f"{base_url}{path}",
        data=json.dumps(payload, ensure_ascii=False).encode("utf-8"),
        headers={
            "content-type": "application/json",
            "x-market-brief-runner-secret": runner_secret,
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(request, timeout=60) as response:  # noqa: S310
            body = response.read().decode("utf-8")
            return json.loads(body) if body else {}
    except urllib.error.HTTPError as exc:
        body = exc.read().decode("utf-8")
        try:
            parsed = json.loads(body) if body else {}
        except json.JSONDecodeError:
            parsed = {}
        raise RuntimeError(parsed.get("error") or f"HTTP {exc.code} calling {path}") from exc
    except urllib.error.URLError as exc:
        raise RuntimeError(f"Network error calling {path}: {safe_error(exc)}") from exc


def fail_job(base_url: str, runner_secret: str, job_id: str, message: str) -> None:
    try:
        post_json(
            base_url,
            "/api/market-briefs/skill-jobs/fail",
            {"job_id": job_id, "error_message": message[:500]},
            runner_secret,
        )
    except Exception as exc:  # noqa: BLE001
        print(f"Failed to mark job as failed: {safe_error(exc)}", file=sys.stderr)


def load_dotenv_if_available() -> None:
    try:
        from dotenv import load_dotenv

        load_dotenv()
    except Exception:
        return


def normalize_base_url(value: str | None) -> str | None:
    if not value:
        return None
    return value.rstrip("/")


def safe_error(exc: Exception) -> str:
    return str(exc).replace("\n", " ")[:500]


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/market-brief-runner/run-market-brief-runner.mjs
#!/usr/bin/env node

const baseUrl = normalizeBaseUrl(process.env.WORKSTATION_BASE_URL);
const runnerSecret = process.env.MARKET_BRIEF_RUNNER_SECRET;
const market = process.env.MARKET_BRIEF_MARKET || "A股";
const runnerName = process.env.MARKET_BRIEF_RUNNER_NAME || "external-skill-runner";

if (!baseUrl) {
  console.error("WORKSTATION_BASE_URL is required, for example http://localhost:3000");
  process.exit(1);
}

if (!runnerSecret) {
  console.error("MARKET_BRIEF_RUNNER_SECRET is required.");
  process.exit(1);
}

const claim = await postJson("/api/market-briefs/skill-jobs/claim", {
  market,
  runner_name: runnerName
});

if (!claim.job_id) {
  console.log(claim.message || "No queued market brief generation jobs.");
  process.exit(0);
}

console.log(`Claimed market brief job ${claim.job_id} for ${claim.market} ${claim.brief_date}.`);

try {
  const generated = buildMockMarketBrief(claim);
  const result = await postJson("/api/market-briefs/skill-result", generated);
  console.log(`Market brief generated: ${result.preview_url}`);
} catch (error) {
  const message = error instanceof Error ? error.message : "External runner failed.";
  await postJson("/api/market-briefs/skill-jobs/fail", {
    job_id: claim.job_id,
    error_message: message
  }).catch(() => null);
  console.error(message);
  process.exit(1);
}

async function postJson(path, payload) {
  const response = await fetch(`${baseUrl}${path}`, {
    method: "POST",
    headers: {
      "content-type": "application/json",
      "x-market-brief-runner-secret": runnerSecret
    },
    body: JSON.stringify(payload)
  });

  const text = await response.text();
  const body = text ? JSON.parse(text) : {};

  if (!response.ok) {
    throw new Error(body.error || `Request failed: ${response.status}`);
  }

  return body;
}

function buildMockMarketBrief(job) {
  const generatedAt = new Date().toISOString();
  const title = `${job.market}市场收评简报｜${job.brief_date}`;
  const summary = "本简报由外部 runner 骨架生成，当前仍使用 mock 数据，用于验证 claim -> result 回写链路。";

  return {
    job_id: job.job_id,
    brief_date: job.brief_date,
    market: job.market,
    title,
    summary,
    markdown_content: [
      `# ${title}`,
      "",
      "> 本简报由 external-skill-runner 骨架生成，当前尚未接入真实行情、新闻源或 AI。",
      "",
      `- 日期：${job.brief_date}`,
      `- 市场：${job.market}`,
      "- 生成方式：external-skill-runner",
      "",
      "## 一、摘要",
      "",
      summary,
      "",
      "## 二、市场概览",
      "",
      "外部 runner 已完成任务领取和回写流程。本段为占位内容，后续 Phase 2L-D-C 可替换为真实行情与新闻摘要。",
      "",
      "## 三、指数表现",
      "",
      "指数表现为 mock 数据，未接入 AkShare、Tushare 或 Wind。",
      "",
      "## 四、风险提示",
      "",
      "本内容仅用于系统联调，不构成投资建议。",
      "",
      "## 数据来源",
      "",
      "- External runner mock fixture",
      "- 当前尚未接入真实行情、新闻或 AI 数据源",
      ""
    ].join("\n"),
    source_snapshot: {
      meta: {
        market: job.market,
        brief_date: job.brief_date,
        runner_name: runnerName,
        generated_at: generatedAt,
        mode: "external_mock",
        data_quality: "mock_only"
      },
      indices: [{ name: "上证指数", code: "000001.SH", close: null, change_pct: null, turnover: null }],
      styles: [],
      sectors: [],
      hot_topics: [],
      capital_flows: [],
      policy_news: [],
      risk_signals: []
    },
    tags: ["市场简报", job.market, "external-runner", "mock", "待复核"],
    data_sources: ["External runner mock fixture", "未接入真实行情数据"]
  };
}

function normalizeBaseUrl(value) {
  if (!value) return null;
  return value.replace(/\/+$/, "");
}
